artichoke-backend/src/extn/core/kernel/integer.rs

```diff
@@ -0,0 +1,177 @@
+    InvalidValue(String, bool),
+    InvalidRadix(String, bool),
+    NoImplicitConversionToString(String, bool),
+}
+
+#[derive(Debug)]
+pub struct Args {
+    pub arg: String,
+    pub radix: Option<i64>,
+    pub raise_exception: bool,
+}
+
+impl Args {
+    pub unsafe fn extract(interp: &Artichoke) -> Result<Self, Error> {
+        let mrb = interp.0.borrow().mrb;
+
+        // TODO: when `base` is not passed, but `exception: true` is,
+        // exception argument goes into arg, which is not right,
+        // we might want to get the argument manually.
+        let (arg, base, exception) = mrb_get_args!(mrb, required = 1, optional = 2);
+
+        let raise_exception: Option<HashMap<String, bool>> = match exception {
+            Some(exception) => match interp.try_convert(Value::new(interp, exception)) {
+                Ok(exception) => exception,
+                Err(_) => None,
+            },
+            _ => None,
+        };
+        let raise_exception: bool = if let Some(raise_exception) = raise_exception {
+            *raise_exception.get("exception").unwrap_or_else(|| &true)
+        } else {
+            true
+        };
+        let radix: Option<i64> = match base {
+            Some(base) => {
+                if let Ok(base) = interp.try_convert(Value::new(interp, base)) {
+                    base
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        };
+
+        let value = Value::new(interp, arg);
+        if value.ruby_type() == Ruby::Fixnum && radix.is_some() {
+            return Err(Error::BaseSpecifiedForNonString(raise_exception));
+        }
+
+        if let Ok(arg) = interp.try_convert(Value::new(interp, arg)) {
+            Ok(Self {
+                arg,
+                radix,
+                raise_exception,
+            })
+        } else {
+            Err(Error::NoImplicitConversionToString(
+                value.ruby_type().class_name().into(),
+                raise_exception,
+            ))
+        }
+    }
+}
+
+pub fn method(interp: &Artichoke, args: &Args) -> Result<Value, Error> {
+    let arg = args.arg.as_str();
+    let radix = args.radix;
+    let raise_exception = args.raise_exception;
+
+    let mut string = String::new();
+
+    // If have mutliple consecutive embedded underscores, argument error!
+    let multi_underscore_re = Regex::new(r"__+").unwrap();
+    if arg.starts_with('_') || arg.ends_with('_') || multi_underscore_re.is_match(arg) {
+        return Err(Error::InvalidValue(arg.into(), raise_exception));
+    }
+
+    // If have multiple consecutive leading/trailing signs, argument error!
+    let multi_sign_re = Regex::new(r"\+\++|\-\-+").unwrap();
+    if multi_sign_re.is_match(arg) {
+        return Err(Error::InvalidValue(arg.into(), raise_exception));
+    }
+
+    // Remove embedded underscore `_`
+    // because they represent error in `from_str_radix` & `from_str`
+    let arg = arg.replace('_', "");
```

Comment:
Passing an `&str` in the `Args` struct would make it clearer that we can replace this replacement and the above `Regex` passes with a single pass over the `&str` for validation and pushing the valid chars into a new mutable `String` buffer.